addons/hw_drivers/tools/upgrade.py

```diff
@@ -0,0 +1,209 @@
+"""Module to manage odoo code upgrades using git"""
+
+import logging
+import platform
+import requests
+import subprocess
+from functools import wraps
+from odoo.addons.hw_drivers.tools.helpers import (
+    odoo_restart,
+    path_file,
+    require_db,
+    toggleable,
+    writable,
+)
+
+_logger = logging.getLogger(__name__)
+
+IS_LINUX = platform.system() == 'Linux'
+
+
+def linux(function):
+    """Decorator to check if the system is Linux before running the function."""
+    @wraps(function)
+    def wrapper(*args, **kwargs):
+        if IS_LINUX:
+            return function(*args, **kwargs)
+    return wrapper
+
+
+def git(*args):
+    """Run a git command with the given arguments, taking system
+    into account.
+
+    :param args: list of arguments to pass to git
+    """
+    git_executable = 'git' if IS_LINUX else path_file('git', 'cmd', 'git.exe')
+    command = [git_executable, f'--work-tree={path_file("odoo")}', f'--git-dir={path_file("odoo", ".git")}', *args]
+
+    p = subprocess.run(command, stdout=subprocess.PIPE, text=True, check=False)
+    if p.returncode == 0:
+        return p.stdout.strip()
+    return None
+
+
+def pip(*args):
+    """Run a pip command with the given arguments, taking system
+    into account.
+
+    :param args: list of arguments to pass to pip
+    """
+    python_executable = [] if IS_LINUX else [path_file('python', 'python.exe'), '-m']
+    command = [*python_executable, 'pip', *args]
+
+    if IS_LINUX and args[0] == 'install':
+        command.append('--user')
+        command.append('--break-system-package')
+
+    p = subprocess.run(command, stdout=subprocess.PIPE, check=False)
+    if p.returncode == 0:
+        return p.stdout.decode().strip()
+    return None
+
+
+def get_db_branch(server_url):
+    """Get the current branch of the database.
+
+    :param server_url: The URL of the connected Odoo database.
+    :return: the current branch of the database
+    """
+    try:
+        response = requests.post(server_url + "/web/webclient/version_info", json={}, timeout=5)
+        response.raise_for_status()
+    except requests.exceptions.HTTPError:
+        _logger.exception('Could not reach configured server to get the Odoo version')
+        return
+    try:
+        return response.json()['result']['server_serie'].replace('~', '-')
+    except ValueError:
+        _logger.exception('Could not load JSON data: Received data is not valid JSON.\nContent:\n%s', response.content)
+        return
+
+
+@toggleable
+@require_db
+def check_git_branch(server_url=None):
+    """Check if the local branch is the same as the connected Odoo DB and
+    checkout to match it if needed.
+
+    :param server_url: The URL of the connected Odoo database (provided by decorator).
+    """
+    db_branch = get_db_branch(server_url)
```

Comment:
this can be None in case of an error, so we could check the value before running git commands